### PR TITLE
[RFR]: Reference input issues

### PR DESCRIPTION
**Pull request by tkvw**
When working with the autosuggest item I encountered several issues with the `ReferenceInput` components, I separated them in this dedicated PR: 
- [X] When setFilter is triggered from child component, the reference and options were retrieved. Only the options should be retrieved (save one roundtrip)
- [X] When the input value of the item is changed, the reference record should be fetched.
- [X] There were alot of rerenders, introduced selector to memoize the referenced records
- [X] The referenced records weren't passed on to the child component, I really think they should be passed on. Introduced `selectedItem` prop in `ReferenceInput` and `selectedItems` prop in `ReferenceArrayInput`. 
- [x] Add tests for new behaviour

### CODE REVIEW

**Comment by djhi** [comment]
Great! It would be nice to include examples in the documentation for `ReferenceInput` and `ReferenceArrayInput` showing how to use the new `selectedItem` and `selectedItems`. I guess the goal here is to create custom inputs with richer content ?

**Comment by tkvw** [comment]
I am looking at this, it looks like you've removed the docs of `ReferenceArrayInput` entirely because of deprecation of `SelectArrayInput`. I've just created a rudimentary implementation of a new `SelectArrayInput` compatible with `material-next` and without dependencies, so it's nice to reference this component when reintroducing the `ReferenceArrayInput` docs.

**Comment by djhi** [comment]
About the `SelectArrayInput`, see https://github.com/TeamWertarbyte/material-ui-chip-input/pull/177

**Comment by djhi** [comment]
About the new `selectedItem`, I guess it can be use to create richer menu items ?

**Comment by tkvw** [comment]
> About the new selectedItem, I guess it can be use to create richer menu items ?

Richer input components. For example, you can use the selectedItem to create an `startAdornment` on the input. This could be achieved by reducing it from the choices, but that's just nonsense when the value is actual available.

**Comment by djhi** [comment]
Can you add an example in the doc ?

**Comment by tkvw** [comment]
I have to come up with one first 😄. `AutocompleteInput` uses it now, reducing the selected item from the choices props is a fallback now.

**Comment by djhi** [comment]
Cool but that does not demonstrate it for our users ;)

**Comment by tkvw** [comment]
> About the SelectArrayInput, see TeamWertarbyte/material-ui-chip-input#177

Without looking at the issue, my implementation of `SelectArrayInput` doesn't use external dependencies.

**Comment by djhi** [comment]
@tkvw bump :)

**Comment by tkvw** [comment]
>  add it to the props 

Yeah, sorry. Not sure what you mean by this exactly

**Comment by djhi** [comment]
I've seen the following warning when trying this PR locally:
> React does not recognize the selectedItem prop on a DOM element.

That's because `selectedItem` is not removed from the `rest` props passed down to children

**Comment by tkvw** [comment]
It should be passed down, just like `choices`, and the childcomponent should take or ignore it. I think you see this because you're using the old `AutocompleteInput`, but I think you're right that `SelectInput` should ignore it as well (and the other components that can be wrapped with `ReferenceInput`)

**Comment by djhi** [comment]
Can you update the `SelectInput` component ?

**Comment by tkvw** [comment]
I'll take a look at it tonight, I'm busy at the moment

**Comment by tkvw** [comment]
Done. Btw: I don't have any problem with committing to my PR, it would save time for these small changes.

File: packages/react-admin/src/mui/input/ReferenceInput.spec.js [hunk 1]
```diff
@@ -225,4 +225,62 @@ describe('<ReferenceInput />', () => {
...
+                crudGetOne={crudGetOne}
+                crudGetMatching={crudGetMatching}
+            >
+                <MyComponent />
+            </ReferenceInput>
+        );
+        expect(crudGetMatching.mock.calls.length).toBe(1);
+
+        wrapper.instance().setFilter('bar');
+        expect(crudGetOne.mock.calls.length).toBe(0);
```
Comment (fzaninotto) [inline packages/react-admin/src/mui/input/ReferenceInput.spec.js]: Same as above, this assertion does not make sense

File: packages/react-admin/src/mui/input/SelectInput.js [hunk 2]
```diff
@@ -36,6 +36,7 @@ const sanitizeRestProps = ({
     reference,
     resource,
     setFilter,
+    selectedItem, // Passed on from ReferenceInput
```
Comment (fzaninotto) [inline packages/react-admin/src/mui/input/SelectInput.js]: IMO, to be removed

File: packages/react-admin/src/reducer/admin/references/possibleValues.js [hunk 3]
```diff
@@ -16,22 +16,16 @@ export default (previousState = initialState, { type, payload, meta }) => {
...
+    possibleValues = possibleValues ? Array.from(possibleValues) : [];
     selectedIds.forEach(
         id =>
             possibleValues.some(value => value == id) ||
             possibleValues.unshift(id)
     );
-
     return possibleValues
-        .map(id => state.admin.resources[reference].data[id])
+        .map(id => state.data[id])
```
Comment (fzaninotto) [inline packages/react-admin/src/reducer/admin/references/possibleValues.js]: what's `state.data`? it doesn't exist.

File: packages/react-admin/src/reducer/admin/references/possibleValues.js [hunk 4]
```diff
@@ -16,22 +16,16 @@ export default (previousState = initialState, { type, payload, meta }) => {
 
 export const getPossibleReferences = (
     state,
-    referenceSource,
-    reference,
+    possibleValues,
```
Comment (fzaninotto) [inline packages/react-admin/src/reducer/admin/references/possibleValues.js]: I don't understand this change. Could you explain why you do it? Could you add a unit test?

File: packages/react-admin/src/mui/input/ReferenceArrayInput.js [hunk 5]
```diff
@@ -126,40 +127,39 @@ export class ReferenceArrayInput extends Component {
...
-            this.fetchReferenceAndOptions();
+            this.fetchReferencesAndOptions();
         }
     };
 
-    fetchReferenceAndOptions(
-        { input, reference, source, resource } = this.props
-    ) {
-        const { pagination, sort, filter } = this.params;
+    fetchReferences = ({ input, reference } = this.props) => {
```
[the remark this answers is not in the document]
Comment (tkvw) [reply packages/react-admin/src/mui/input/ReferenceArrayInput.js]: They are never called separately, but they **should be**. When rendering an `autocompleteinput` the `selectedItem` is retrieved on initial mount. As long as the selectedItem does not change, there's no need to fetch the item over and over again while the `searchText` changes. The references should be retrieved on each change of the searchText. So it eliminates unnecessary fetches. 
So when `setFilter` is called, only the references should be fetched and when the value changes the `getOne` should be called.

File: docs/Inputs.md [hunk 6]
```diff
@@ -509,7 +509,7 @@ import { RadioButtonGroupInput, ReferenceInput } from 'react-admin'
 
 ## `<ReferenceInput>`
 
-Use `<ReferenceInput>` for foreign-key values, i.e. to let users choose a value from another REST endpoint. This component fetches the possible values in the reference resource (using the `GET_LIST` REST method), then delegates rendering to a subcomponent, to which it passes the possible choices as the `choices` attribute.
+Use `<ReferenceInput>` for foreign-key values, i.e. to let users choose a value from another REST endpoint. This component fetches the possible values in the reference resource (using the `GET_LIST` REST method) and the referenced record (using the `GET_ONE` REST method), then delegates rendering to a subcomponent, to which it passes the possible choices as the `choices` attribute and the referenced record as `selectedItem`.
```
[the remark this answers is not in the document]
Comment (tkvw) [reply docs/Inputs.md]: The value is available and it saves computation, so I don't understand why it isn't passed on. I think relying on a `selectedItem` being passed on vs looking up the value yourself using the choices array just makes more sense imo.

File: packages/react-admin/src/mui/input/ReferenceArrayInput.spec.js [hunk 7]
```diff
@@ -194,4 +194,62 @@ describe('<ReferenceArrayInput />', () => {
...
+                crudGetMany={crudGetMany}
+                crudGetMatching={crudGetMatching}
+            >
+                <MyComponent />
+            </ReferenceArrayInput>
+        );
+        expect(crudGetMatching.mock.calls.length).toBe(1);
+
+        wrapper.instance().setFilter('bar');
+        expect(crudGetMany.mock.calls.length).toBe(0);
```
[the remark this answers is not in the document]
Comment (tkvw) [reply packages/react-admin/src/mui/input/ReferenceArrayInput.spec.js]: The assertion is wrong indeed, will fix it

File: packages/react-admin/src/reducer/admin/references/possibleValues.js (hunk 4 above)
Comment (tkvw) [reply packages/react-admin/src/reducer/admin/references/possibleValues.js]: `getPossibleReferences` no longer receives the entire state, but a selection of the state retrieved by a memoized selector in `ReferenceInput`. This safes unnecessary rerenders.

File: packages/react-admin/src/reducer/admin/references/possibleValues.js (hunk 3 above)
Comment (tkvw) [reply packages/react-admin/src/reducer/admin/references/possibleValues.js]: it's a partial state received from the reselect selector created in `ReferenceInput`

File: packages/react-admin/src/mui/input/ReferenceArrayInput.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (tkvw) [reply packages/react-admin/src/mui/input/ReferenceArrayInput.js]: Oh, wow, I see now what you mean, I didn't change it correctly. `ReferenceInput` does make the calls separately. I'll need to fix this.

[the remark this answers is not in the document]
Comment (tkvw) [reply packages/react-admin/src/mui/input/ReferenceArrayInput.js]: This is fixed in https://github.com/marmelab/admin-on-rest/pull/1420/commits/c73142c9d0bd5ec51962ffd185ce997a6be81f20, brought the behaviour inline with `ReferenceInput`

File: docs/Inputs.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (fzaninotto) [reply docs/Inputs.md]: You didn't understand me. This value is already passed in the form of the `choices` and `input.value`, so you have to give me a proper use case for passing it twice. It's not needed in `SelectInput` or `AutoCompleteInput`, and I don't understand why it would be needed in any other similar component. 

I'm ready to change my opinion about this if you provide a strong use cas for it. But as it delays the merge of this PR, I strongly suggest that you remove all `selectedItem` related code from this PR.

File: packages/react-admin/src/mui/input/ReferenceArrayInput.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (fzaninotto) [reply packages/react-admin/src/mui/input/ReferenceArrayInput.js]: Now it's OK, thanks.

File: packages/react-admin/src/mui/input/ReferenceArrayInput.js [hunk 8]
```diff
@@ -276,26 +286,31 @@ ReferenceArrayInput.defaultProps = {
     perPage: 25,
     sort: { field: 'id', order: 'DESC' },
     referenceRecords: [],
+    record: {},
```
Comment (fzaninotto) [inline packages/react-admin/src/mui/input/ReferenceArrayInput.js]: why this change? I'm afraid it may have side effects.

File: packages/react-admin/src/reducer/admin/references/possibleValues.js (hunk 3 above)
Comment (fzaninotto) [reply packages/react-admin/src/reducer/admin/references/possibleValues.js]: OK, then don't use the name `state`, it's misleading.

File: packages/react-admin/src/reducer/admin/references/possibleValues.js (hunk 4 above)
Comment (fzaninotto) [reply packages/react-admin/src/reducer/admin/references/possibleValues.js]: then rename the `state` param to something more significant.

File: docs/Inputs.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (tkvw) [reply docs/Inputs.md]: Yes I understand, here goes my strong usecase 😄;  

Consider `ReferenceInput` wrapping `AutocompleteInput` and a resource is referenced with 100.000 records, you don't want to retrieve the entire resultset, so you'll return a limited subset (the first 25 or so) so the user has something to select, but if this subset does not include the record of `input.value` the selectedItem can't be resolved from the choices! It's possible to always include the selectedItem in the choices, by changing the selector; but then the user will always see the current value in the selection items and this could lead to unexpected results. For example the current input value is resolved to "France" and the user types "Nether" and the dropdown still shows "France" as viable option. 

This is solved by passing through `selectedItem`. Btw; this only makes sense for `Autocomplete`, because for SelectInput all records of the resource should be resolved.

[the remark this answers is not in the document]
Comment (fzaninotto) [reply docs/Inputs.md]: This case is already handled by the `getPossibleReferences` selector in the `references/possibleValues` reducer. So the selected item is always in the choices, whatever the number of actual choices. 

Do you have another valid use case?

[the remark this answers is not in the document]
Comment (tkvw) [reply docs/Inputs.md]: Last attempt; it does handle it, but I am arguing it has an unwanted side effect. It will always show the current value in the suggestion list. I think this is counter intuitive. 

This example shows I am looking for "Illum veri", but I get the current value in the suggestion list. Filtering the choices list (to remove input.value) before showing the suggestions isn't an option either, because if I type "Perspi" it should be included in the choices list. 

If this doesn't convince you, I'll remove it from the PR.

[the remark this answers is not in the document]
Comment (fzaninotto) [reply docs/Inputs.md]: I agree that it may be counterintuitive, but both the SelectInput and the AutocompleteInput need the current value to be in the choices in their current implementation. So to remove the current choice in the GUI, you need to change not only the ReferenceItem component, but also its potential child components. Changing these components is very not BC, and IMO way beyond the purpose of this PR. 

So no, I’m not convinced.

[the remark this answers is not in the document]
Comment (tkvw) [reply docs/Inputs.md]: Ok, let's put a pin in it. No trace of `selectedItem` in the PR anymore. https://github.com/marmelab/admin-on-rest/pull/1420/commits/14fb0e8568da94b83925383b74645204757bb638

[the remark this answers is not in the document]
Comment (fzaninotto) [reply docs/Inputs.md]: There are still traces of it, and there are still comments that you haven't addressed. Please review your own PR using the "Files changed" tab, since the discussion tab is too long to be useful (another good reason to make small and atomic PRs by the way).

**Comment by fzaninotto** [comment]
needs rebase

File: packages/react-admin/src/mui/input/ReferenceArrayInput.js (hunk 8 above)
Comment (tkvw) [reply packages/react-admin/src/mui/input/ReferenceArrayInput.js]: The component expects a record in `componentWillReceiveProps` and it's not required, so a default value made sense to me, not sure about what side effects you're thinking about, but I can remove it in favor of extra checks in `componentWillReceiveProps` or making the prop required.

**Comment by fzaninotto** [comment]
Thanks!
